### PR TITLE
Density Matrix in Python

### PR DESCRIPTION
**Pull request by gcarleo**
I am porting also density matrices to python (basically, the last big thing to be ported from c++....).

It turns out that the (super ugly) doubling of the hilbert space needed in c++ was stemming mostly from the limitations of the sampler interface, that were depending on hilbert to produce initial random states and to deduce the size of the input. 
Since these are distinct concepts, I implemented things in a way that general samplers never see or know anything about hilbert. The just know about the ```input_size``` of the machines they are dealing with, and they have a ```kernel``` that does two things: initialize a random state, and propose a transition. 

This basically allows to define specific (small) kernels that are specifically suited for density matrices, or just use generic kernels (like MetropolisLocal that work in both cases). 

For the moment I also drafted a version of ```AbstractDensityMatrix```, @PhilipVinc  please let me know if it makes sense. The only potential issue is how to get a view (not a copy) when calling .diagonal() ...

### CODE REVIEW

**Comment by gcarleo** [comment]
@PhilipVinc now this works (without doubling hilbert spaces): 

```python
import netket as nk
import numpy as np

g = nk.graph.Hypercube(length=4, n_dim=1)

# Hilbert space of spins on the graph
hi = nk.hilbert.Spin(s=0.5, graph=g)


# RBM Spin Machine
ma = nk.machine.density_matrix.RbmSpin(hilbert=hi, alpha=1)
ma.init_random_parameters(seed=1234, sigma=0.001)

print(np.trace(ma.to_matrix(normalize=True)))

# Metropolis Local Sampling
sa = nk.sampler.MetropolisLocal(machine=ma, n_chains=2)

samples = sa.generate_samples(10)
print(samples.shape)

# Sampler for the diagonal of the density matrix, to compute observables
sa_obs = nk.sampler.MetropolisLocal(machine=ma.diagonal(), n_chains=2)

samples = sa_obs.generate_samples(10)
print(samples.shape)
```

The Rbm here is just a Rbm with twice the number of visible units, so it does not have the purification constraints on the weights. It was just for testing... (also notice that how I handle symmetries shouldn't be correct, most likely I am missing a conjugate, but it's a proof of principle).

Can you test these things and see if the design is correct against the steady state solver? If so, we should then just write a purified Rbm...

**Comment by gcarleo** [comment]
I have also implemented the complex-valued purified Rbm in Jax. 
This now works as expected:

```python
import netket as nk
import numpy as np

g = nk.graph.Hypercube(length=4, n_dim=1)

# Hilbert space of spins on the graph
hi = nk.hilbert.Spin(s=0.5, graph=g)

ma = nk.machine.density_matrix.JaxNdmSpin(hilbert=hi, alpha=1, beta=1)
ma.init_random_parameters()

assert ma.input_size == hi.size * 2


def check(a, rtol=1e-05, atol=1e-08):
    # Hermitean
    assert np.allclose(a, a.transpose().conjugate(), rtol=rtol, atol=atol)

    # Positive definite
    eigvals = np.linalg.eigvalsh(a)
    assert np.min(eigvals) >= 0

    # Trace 1
    assert np.allclose(np.trace(a), 1, rtol=rtol, atol=atol)
    print("all good")


mat = ma.to_matrix(normalize=True)
check(mat)

# Metropolis Local Sampling
sa = nk.sampler.jax.MetropolisLocal(machine=ma, n_chains=2)

samples = sa.generate_samples(10)
print(samples.shape)


# Metropolis Local Sampling of the Diagonal
sa = nk.sampler.jax.MetropolisLocal(machine=ma.diagonal(), n_chains=2)

samples = sa.generate_samples(10)
print(samples.shape)
```

@PhilipVinc I leave the rest to you, especially testing/fixing this within the steady state solver and also removing the C++ code when you are done! Thanks

**Comment by PhilipVinc** [comment]
I'm on it.

**Comment by PhilipVinc** [comment]
The main problem in getting those things to work with Jax is that, as it is implemented now, I need to compute the gradient of c_loc, which cannot be written as a vjp product...

**Comment by PhilipVinc** [comment]
Just for a reminder:
 - Jax has a problem doing backward differentiation on R-> C functions (essentially he does the forward pass, then back propagates a real unit instead of a complex one...). The solution I found at the moment is to convert the weights to complex when computing the `der_log` of a real-valued jax machine. You also must specify `holomorphic=True`. We should raise this issue upstream.

 - I have changed stats/mean in order to work with Jax/MPI or whatever other thing we have on hand, by exploiting `sum_inplace`. please check that this is ok.

 - `der_log_vals` now is special-cased for jax. In that case, we are doing the work twice (we compute log_vals and der_log_vals), and we could de-duplicate this.

**Comment by PhilipVinc** [comment]
Among other things:
 - The PR by @chrisrothUT added an export to `vmc_common.h`, that made it so that if jax was present, their `tree_map` would be exported, shadowing ours. I reverted this change, because I feel that this might cause some hard-to-find bugs if we start to have methods that are overridden sometimes. Better to be sure our implemntation works all the time, and if someone needs the jax one, he can import it explicitly

**Comment by PhilipVinc** [comment]
MPI tests are failing because I removed the out keyword from `mean`.

In short, we should stop using out if possible, because in-place modifications of buffers are not jax-compatible... 
Though I agree that this is not nice, because things like numpy would benefit (somewhat) from it. 

I believe that the allocation cost of the reduced (so, smaller) output array should be somewhat negligible in big systems, so I would go forward with it.
Do you have anything against it? @gcarleo @femtobit

**Comment by gcarleo** [comment]
@PhilipVinc you mean that sum_inplace basically is being removed?

**Comment by gcarleo** [comment]
Nevermind, I see what you mean now. In practice you want to remove these 
```python
out = _np.asarray(_np.mean(a, axis=axis, out=out))
```
because you want to make things backend agnostic... I guess it makes sense, I don't have a strong opinion in favor of out

**Comment by PhilipVinc** [comment]
No...

In an effort to make `stats.mean` work with any array type (jax et al) I made the following change, using _sum_inplace.

old:
```python
def mean(a, axis=None, out=None):		
    out = _np.asarray(_np.mean(a, axis=axis, out=out))		
    _MPI_comm.Allreduce(MPI.IN_PLACE, out.reshape(-1), op=MPI.SUM)		
    out /= _n_nodes		
    return out		
```

new:
```python
def mean(a, axis=None):		
    out = a.mean(axis=axis)
    _sum_inplace(out)
    out /= _n_nodes		
    return out		

```

The problem is that if we want to be general, we cannot use `_np.asarray()`. 
But if we reduce across all axes, mean of a numpy array returns a scalar, which cannot be summed inplace by MPI.

So... I think we should cheat, and we make `_sum_inplace` always work in place except for scalars,

**Comment by PhilipVinc** [comment]
The main point is also that to make `mean` work with jax, we cannot pass an `out=` argument, so the two things go well together

**Comment by gcarleo** [comment]
Yes I understand, I also think that it's the same for Torch tensors

**Comment by PhilipVinc** [comment]
Good. Actually I think we could do the same also for `sum`.

We could even keep the behaviour of non-allocating out in the case of numpy if we wanted with this trick:
```
def sum(a, **kwargs):
    out = a.sum(**kwargs)
    out = _sum_inplace(out)
    
    return out
```

File: netket/machine/jax.py [hunk 1]
```diff
@@ -168,7 +165,7 @@ def vector_jacobian_prod(
 
     @property
     def is_holomorphic(self):
-        return self._dtype is complex
+        return True
```
Comment (gcarleo) [inline netket/machine/jax.py:168]: I think that this shouldn't be changed (basically because it is used by SR to determine the kind of update), but instead we should always use True in the jacobian only?

File: netket/machine/jax.py [hunk 2]
```diff
@@ -51,7 +51,9 @@ def __init__(self, hilbert, module, dtype=complex):
         self._forward_fn = jax.jit(self._forward_fn)
 
         forward_scalar = jax.jit(lambda pars, x: self._forward_fn(pars, x).reshape(()))
-        grad_fun = jax.jit(jax.grad(forward_scalar, holomorphic=self.is_holomorphic))
+
+        # FIXME: not everything is holomorphic.
+        grad_fun = jax.jit(jax.grad(forward_scalar, holomorphic=true))
```
Comment (gcarleo) [inline netket/machine/jax.py:56]: capital True

Comment (gcarleo) [inline netket/machine/jax.py:56]:
```suggestion
        grad_fun = jax.jit(jax.grad(forward_scalar, holomorphic=True))
```

**Comment by PhilipVinc** [comment]
this black thing is evil
